feat: New methods to ingest and create Feature Groups

JoseJuan98 opened the pull request:
*Issue #, if available:*

*Description of changes:* added a new methods to automate processes to work with Feature Groups, one to prepare a FG before creation and another to extract directly a dataframe from a FG faster. Most relevant methods:

- `get_feature_group_as_dataframe(...)`: method that automates the extraction of the data in a FeatureGroup selecting by version, by latest ingestion or by all data ingested.

-  `prepare_fg_from_dataframe_or_file(...)`: method that automates all the logic to prepare the new batch data to be ingested into a FeatureGroup.

Also, I extended the method `FeatureGroup.as_dataframe()` to accept kwargs to specify some parameters in the `pandas.read_csv` use to create the dataframe as I encounter some problems with encoding and dtypes that could be fix specifying `low_memory=False` and the encoding used.

*Testing done:*
Local tests with tox following the contribution guide.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

JoseJuan98 commented:
Hi @claytonparnell I open this PR because I had one previously that I didn't rebase and it was messy now to rebase, so I closed it and open this one. I have check the tox tests (sphinx,flake8,pylint,docstyle,black-check,twine) and they pass succesfully and also I have check the unit tests and the integration tests in my AWS account.

Could you please run the bot to run the tests?

navinsoni reviewed (commented):
/bot run all

navinsoni commented on src/sagemaker/feature_store/feature_utils.py:
```diff
@@ -0,0 +1,334 @@
+
+        metadata = sts.assume_role(RoleArn=assume_role, RoleSessionName="SagemakerExecution")
+
+        access_key_id = metadata["Credentials"]["AccessKeyId"]
```
use `.get` method here to prevent exception if Credentials is not present in metadata.

JoseJuan98 replied:
done

navinsoni commented on src/sagemaker/feature_store/feature_utils.py:
```diff
@@ -0,0 +1,334 @@
+    if session is not None:
+        sagemaker_session = session
+    elif region is not None:
+        sagemaker_session = get_session_from_role(region=region, assume_role=role)
```
what happens if role is `None` here

JoseJuan98 replied:
If role is `None` it will use the execution role by default assigned to the sagemaker domain in that region

JoseJuan98 reviewed (commented):
Made changes

JoseJuan98 commented on src/sagemaker/feature_store/feature_utils.py:
```diff
@@ -0,0 +1,334 @@
+    # It will try to assume the role specified
+    if assume_role:
+        sts = boto_session.client(
+            "sts", region_name=region, endpoint_url="https://sts.eu-west-1.amazonaws.com"
```
Okay, I will do that, thank you for the feedback :)

JoseJuan98 commented:
Hi @navinsoni  , I made the changes requested, could you run test again and review them to check everything is in order?

navinsoni commented:
> Hi @navinsoni , I made the changes requested, could you run test again and review them to check everything is in order?

starting.

navinsoni reviewed (commented):
/bot run all

claytonparnell commented:
@JoseJuan98 please run `tox -e black-format` to resolve formatting issues

JoseJuan98 commented:
> @JoseJuan98 please run `tox -e black-format` to resolve formatting issues

@navinsoni  and @claytonparnell, I fixed it, and in the integration tests I guess I don't understand what the fixture `region_name` returns, it returns `feature_store_session.boto_session.region_name`, in the docs is shown as `Any` and my method need and string, that's why is failing. 
I changed it now to `str(region_name)`, if you have any suggestions please let me know, if everything is alright would you mind rerunning the tests, please?

navinsoni reviewed (commented):
/bot run all

JoseJuan98 commented:
Hi @navinsoni  and @claytonparnell , I fixed the end of the line missing in the docs and other error in the tests. But I get the following error in a part of the code I didn't change:

```
sagemaker.exceptions.UnexpectedStatusException: Error hosting endpoint model-quality-monitor-integ-1678733175-df96: 
Failed. Reason: The primary container for production variant AllTraffic did not pass the 
ping health check. Please check CloudWatch logs for this endpoint..
```
In `tests/integ/test_model_quality_monitor.py:89`

claytonparnell commented:
@JoseJuan98 That happens intermittently when there's an issue with the endpoint during testing, doesn't look like that error is related to your changes. I'll restart the tests

claytonparnell reviewed (commented):
/bot run all

JoseJuan98 commented:
> Please fix this: `test_athena_query_as_dataframe` `E Expected: read_csv('tmp/query_id.csv', delimiter=',')` `E Actual: read_csv(filepath_or_buffer='tmp/query_id.csv', delimiter=',')`

fixed and tested, could you run tests again?

JoseJuan98 commented:
@navinsoni, @alex-tang, @claytonparnell could you please run the test?

JoseJuan98 commented:
/bot run all

jpenagom commented:
@navinsoni, @alex-tang, @claytonparnell could you please run CodeBuild? Or can it be run be me?

JoseJuan98 commented:
@navinsoni , @alex-tang , @claytonparnell  could you please run CodeBuild?

JoseJuan98 commented:
@navinsoni , @alex-tang  , @claytonparnell  any updates of the status of this PR? could you please run CodeBuild?

JoseJuan98 commented:
It has been 6 weeks without any comment or check, please could anyone of @vikas0203 , @navinsoni , @alex-tang , @claytonparnell give any updates on the status of this PR? Also, could anyone of you please run the CodeBuild to check the integration tests?

knikure reviewed (commented):
/bot run all

JoseJuan98 commented:
Hi @knikure , thank you for executing the tests. I got an error in a [test](https://devbuildsagemakerpythonsdkstackpdx-sage-buildlogs-5gc8kag9othm.s3.amazonaws.com/6f413466-825d-4f0f-a033-97cb1b192ca7/build.log?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIASCMSNBKCUK6XCE4H%2F20230517%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20230517T072523Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Security-Token=IQoJb3JpZ2luX2VjEBgaCXVzLXdlc3QtMiJIMEYCIQDdi3jujrtvR8TCiWlMewAGSaXy7x58CY8oBgrcdvOz3gIhAM1ovMlFlX6T91tD9hory6KmKTKE7geJgpxd5FjC%2Fb8XKtUDCEEQAxoMMTQyNTc3ODMwNTMzIgzKKqgh7fxb9O19vWYqsgPe16AqzusisqiG%2BUYKyL0p5yVZkl4M3j%2F9cWBUxjnzf32O%2FzBgY%2BWAIhqi3St56s6JU5uGCrEWPwMt7XHGyOQUDDj22JzRv84wRPKBvTs8%2Bmkla4AsxLf7eVZpRdaJTjZEBQRwOcgdOBd4ctjk2Z5fTj6CPtsPUIF3B4jDrumsWaXK8e78h4EgyAZ2939tnbuQN47TvISYUg83kKrNOudjU0QbAkYcq29BogXNkNtPrBAtOJIHvhT%2F4OsM8NTHi7IBZBVPkbxvIX7Xr1UffrNA%2FS3U%2BU2OxG1k%2BAs0QnufXg9wwXTHUbtBgotgInrrup%2BmTgwNJA2f%2B%2F1hrCyS%2Fob4W9Apu4D7Uk%2BqYPEwfUxGKM%2FDrrSPfUnzw%2FSEEPH3j9O8lGfCj3jemR6z%2BUy1a0BsxZ0P%2BinxHYpCiRW7u9tAJrNftr41PooD%2FLIQBIrhHlysnowsGLID%2FjpeEaUep8Iq826pJ4b48ySK333m5cs3DEeg0rXECLk1bVMKSpW5pwr2eiNRLbmBndNtGy2vXuar0MWQCIuUkXBrqUlhApZbU6Ow9rz06KoQDq59OqOGl7cYejDigpKjBjqdAaWsdjqFKwZUhH0twjrU0C0TDt6uGhFjRlj7JraIexPnaccF2ohxV9chTKzJ9PTxGs0zZ9A5SLM5ceEYq%2FSBqUqAF%2BP3yHERjzqY9nLCPgKyQEXUb0IxRVAPVOL1BzlPQYMEtWTy9KvOBmJuQwGB2rrcmfTu1GpssGNXWTNMxbmjjP7urFebBOcZmfb%2FjiQg22GdgJJL0BGBjgmnlsY%3D&X-Amz-Signature=84962050ecd12e21c0d509a9aaf2457b9dd37d480bff786f4782aafd7bc42396) not related with my changes, do you know why it could be that or how to fix it?

```
==================================== ERRORS ====================================
_____________ ERROR collecting tests/integ/test_airflow_config.py ______________
tests/integ/test_airflow_config.py:53: in <module>
    seconds_to_sleep=6,
.tox/py37/lib/python3.7/site-packages/sagemaker/utils.py:672: in retries
    exception_message_prefix, max_retry_count
E   Exception: 'airflow import ' has reached the maximum retry count of 10
------------------------------- Captured stdout --------------------------------
Received: Unable to configure formatter 'airflow'
Received: Unable to configure formatter 'airflow'
Received: Unable to configure formatter 'airflow'
Received: Unable to configure formatter 'airflow'
Received: Unable to configure formatter 'airflow'
Received: Unable to configure formatter 'airflow'
Received: Unable to configure formatter 'airflow'
Received: Unable to configure formatter 'airflow'
Received: Unable to configure formatter 'airflow'
_____________________________ ERROR collecting gw4 _____________________________
Different tests were collected between gw0 and gw4. The difference is:
--- gw0

+++ gw4

@@ -1,10 +1,10 @@
```

knikure reviewed (commented):
/bot run all

JoseJuan98 commented:
> Hi @knikure , I get the following error related with the name of the version:
> 
> ```
> ERROR: For req: sagemaker==2.156.1.dev0. sagemaker has an invalid wheel, could not read 'sagemaker-2.156.1.dev0.dist-info/WHEEL' file: KeyError("There is no item named 'sagemaker-2.156.1.dev0.dist-info/WHEEL' in the archive")
> WARNING: You are using pip version 21.3; however, version 23.1.2 is available.
> You should consider upgrading via the '/codebuild/output/src698592327/src/github.com/aws/sagemaker-python-sdk/.tox/pylint/bin/python -m pip install --upgrade pip' command.
> ```
> 
> Shall I change it to one accepted by the wheel format? Maybe to "2.156.1" ? Sorry, but I don't if you follow any convention for naming versions, but this one seems to fail.

@knikure I changed the version to 2.156.1 to be a valid format supported by wheel.

JoseJuan98 commented:
@knikure could you run the tests again, please?

knikure reviewed (commented):
/bot run all

JoseJuan98 commented:
> > @knikure could you run the tests again, please?
> 
> done! Can you change the name of the PR with relevant information with feature:

done! Is it better now?

JoseJuan98 commented:
@knikure the tests fails because of the following error in a script that I didn't modify, do you know why?

```shell
==================================== ERRORS ====================================
_____________ ERROR collecting tests/integ/test_airflow_config.py ______________
tests/integ/test_airflow_config.py:53: in <module>
    seconds_to_sleep=6,
.tox/py37/lib/python3.7/site-packages/sagemaker/utils.py:675: in retries
    exception_message_prefix, max_retry_count
E   Exception: 'airflow import ' has reached the maximum retry count of 10
------------------------------- Captured stdout --------------------------------
Received: Unable to configure formatter 'airflow'
Received: Unable to configure formatter 'airflow'
```

knikure reviewed (commented):
/bot run all

JoseJuan98 commented:
@knikure I think fixed run notebook tests that failed, could you run them all again please?

knikure reviewed (commented):
/bot run all

JoseJuan98 commented:
Hi @knikure , any idea why the following errors - not related to my changes - in  ` sagemaker-python-sdk-pr` ?

```
E               sagemaker.exceptions.CapacityError: Error for Training job huggingface-pytorch-training-
2023-06-13-16-26-16-386: Failed. Reason: CapacityError: Unable to provision requested ML compute capacity. 
Please retry using a different ML instance type.
```


And also in `sagemaker-python-sdk-notebook-tests`:

```
Caused by: java.lang.IllegalStateException: Notebook Instance samplenotebooktest-p-pdx-10-1686674016675 entered 
unexpected terminal state Failed. (FailureReason: Notebook Instance Lifecycle Config 'arn:aws:sagemaker:us-
west-2:142577830533:notebook-instance-lifecycle-config/install-python-sdk-4f88d2ea' for Notebook Instance 
'arn:aws:sagemaker:us-west-2:142577830533:notebook-instance/samplenotebooktest-p-pdx-10-1686674016675' took 
longer than 5 minutes. Please check your CloudWatch logs for more details if your Notebook Instance has Internet 
access.)
```

JoseJuan98 commented:
Hi @knikure , could you run tests?

JoseJuan98 commented:
Hi @knikure @alex-tang @claytonparnell @navinsoni , could you run the tests, please?

Is there anything missing in this PR a part from passing the tests? This PR has been open since November (considering the older PR request that was open) and I didn't receive much information from you since them, if you think this PR is not useful I can just close it.

knikure reviewed (commented):
/bot run all

JoseJuan98 commented:
@knikure the branch was not updated with the latest master branch, could you run them again please?

JoseJuan98 commented:
@knikure could you execute the tests? Is there anything missing in this PR a part from passing the tests?

JoseJuan98 commented:
> When is going to be available this feature?

This PR needs to be approved and most of test were passing only ones non dependent in my changes were failing, I guess we need to ask @knikure, @alex-tang , @navinsoni or @claytonparnell .

This PR has been open for a very long time with barely any feedback, I am considering closing it if I will not get any feedback soon.

JoseJuan98 reviewed (commented):
Requested changes were made

JoseJuan98 commented:
> I guess we need to ask @knikure, @alex-tang , @navinsoni or @claytonparnell .
> 
> This PR has been open for a very long time with barely any feedback, I am considering closing it if I will not get any feedback soon.

@knikure, @alex-tang , @navinsoni or @claytonparnell this PR has been open for a very long time with barely any feedback, it seems useful work for people using Feature Stores, but I guess I should close unless any of you review it and tell me otherwise.

knikure commented:
> > I guess we need to ask @knikure, @alex-tang , @navinsoni or @claytonparnell .
> > This PR has been open for a very long time with barely any feedback, I am considering closing it if I will not get any feedback soon.
> 
> @knikure, @alex-tang , @navinsoni or @claytonparnell this PR has been open for a very long time with barely any feedback, it seems useful work for people using Feature Stores, but I guess I should close it unless any of you review it and tell me otherwise.

@JoseJuan98 Thanks for your contribution and sorry for the delayed response. Let me take a look and get back to you.

knikure reviewed (commented):
/bot run notebook-tests

knikure commented on tests/scripts/run-notebook-test.sh:
```diff
@@ -75,7 +75,7 @@ for env in base /home/ec2-user/anaconda3/envs/*; do
     $ENV_PYTHON -m pip install "$TARBALL_DIRECTORY/sagemaker.tar.gz"
 
-    sudo -u ec2-user -E sh -c 'source /home/ec2-user/anaconda3/bin/deactivate'
+    sudo -u ec2-user -E sh -c 'conda deactivate'
```
I believe notebbok-tests are failing due to this? Can you revert this change?

JoseJuan98 replied:
Done. Could you execute tests again?

benieric reviewed (commented):
/bot run pr

benieric reviewed (commented):
/bot run all

JoseJuan98 commented:
hi @benieric , I see all the checks have passed. Do you know the next steps to follow to merge the PR? Please let me know if there is anything required from my side.

Thank you,
Jose.

benieric commented:
@JoseJuan98 Just need to get approval from a reviewer and address any comments they may have.
